spring-graphql/src/main/java/org/springframework/graphql/data/method/annotation/support/AuthenticationPrincipalArgumentResolver.java

```diff
@@ -0,0 +1,145 @@
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.expression.BeanResolver;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ExpressionParser;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
+import org.springframework.graphql.data.method.HandlerMethodArgumentResolver;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.security.core.context.ReactiveSecurityContextHolder;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+import reactor.core.publisher.Mono;
+
+import java.lang.annotation.Annotation;
+
+/**
+ * Resolver to obtain {@link Authentication#getPrincipal()} from Spring Security context via
+ * {@link SecurityContext#getAuthentication()} for parameters annotated with {@link AuthenticationPrincipal}.
+ *
+ * <p>The resolver checks both ThreadLocal context via {@link SecurityContextHolder}
+ * for Spring MVC applications, and {@link ReactiveSecurityContextHolder} for
+ * Spring WebFlux applications.
+ *
+ * @author Rob Winch
+ * @since 1.0.0
+ */
+public class AuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
+
+	private ExpressionParser parser = new SpelExpressionParser();
+
+	private BeanResolver beanResolver;
+
+	/**
+	 * Sets the {@link BeanResolver} to be used on the expressions
+	 * @param beanResolver the {@link BeanResolver} to use
+	 */
+	public void setBeanResolver(BeanResolver beanResolver) {
+		this.beanResolver = beanResolver;
+	}
+
+	@Override
+	public boolean supportsParameter(MethodParameter parameter) {
+		return findMethodAnnotation(AuthenticationPrincipal.class, parameter) != null;
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, DataFetchingEnvironment environment) throws Exception {
+		return getCurrentAuthentication().map(Authentication::getPrincipal)
+				.map((principal) -> resolvePrincipal(parameter, principal))
+				.transform((argument) -> {
+					Class<?> parameterType = parameter.getParameterType();
+					boolean isParameterPublisher = Publisher.class.isAssignableFrom(parameterType);
+					return isParameterPublisher ? Mono.just(argument) : argument;
+				});
+	}
+
+	private Mono<Authentication> getCurrentAuthentication() {
+		SecurityContext securityContext = SecurityContextHolder.getContext();
+		return Mono.justOrEmpty(securityContext.getAuthentication())
+				.switchIfEmpty(ReactiveSecurityContextHolder.getContext().map(SecurityContext::getAuthentication));
+	}
+
+	private Object resolvePrincipal(MethodParameter parameter, Object principal) {
```

Comment:
This method can return `null` but is not marked `@Nullable`. A `null` return value would be an issue for the `map` operator. I'm not actually sure what the expected outcome is if the resolved user type does not match the declared method parameter type and errorOnInvalidType is false? Wouldn't it be a ClassCastException anyway, or is the idea to pass `null` or empty Mono? The Javadoc doesn't say what happens but I don't see any other outcomes..

Reply:
This has been resolved.